File: ReferenceDesigns/w3_802.11/wlan_exp/warpnet/classes/wn_transport_eth_udp.m
%******************************************************************************
% WARPNet Ethernet UDP Transport
%
% File   :	wn_transport_eth_udp.m
% Authors:	Chris Hunter (chunter [at] mangocomm.com)
%			Patrick Murphy (murphpo [at] mangocomm.com)
%           Erik Welsh (welsh [at] mangocomm.com)
% License:	Copyright 2013, Mango Communications. All rights reserved.
%			Distributed under the WARP license  (http://warpproject.org/license)
%
%******************************************************************************
% MODIFICATION HISTORY:
%
% Ver   Who  Date     Changes
% ----- ---- -------- -------------------------------------------------------
%
%******************************************************************************

classdef wn_transport_eth_udp < wn_transport & handle_light
% Ethernet UDP Transport parent class for unicast traffic
% User code should not use this object directly-- the parent wn_node will
%  instantiate the appropriate transport object for the hardware in use

%******************************** Properties **********************************

    properties (SetAccess = public)
        timeout;        % Maximum time spent waiting before retransmission
    end

    properties (SetAccess = protected, Hidden = true)
        type;           % WARPNet transport type
        sock;           % UDP socket
        status;         % Status of UDP socket
        maxPayload;     % Maximum payload size (e.g. MTU - ETH/IP/UDP headers)
    end

    properties (SetAccess = public)
        eth_MAC_addr;   % Node's MAC address
        address;        % IP address of destination
        port;           % Unicast port of destination
        bcast_port;     % Broadcast port of destination
        hdr;            % Transport header object
        group_id;       % Group ID of the node attached to the transport
        rxBufferSize;   % OS's receive buffer size in bytes
    end

    properties(Hidden = true,Constant = true)
        %These constants define specific command IDs used by this object.
        %Their C counterparts are found in wn_transport.h
        GRP = 'transport';
        CMD_PING            = 1;
        CMD_IP              = 2;
        CMD_PORT            = 3;
        CMD_PAYLOADSIZETEST = 4;
        CMD_NODEGRPID_ADD   = 5;
        CMD_NODEGRPID_CLEAR = 6;        
    end
    
    properties(Hidden = true,Constant = true)
        % These constants define the specific parameter identifiers used by this object.
        % All nodes must implement these parameters but can be extended to suppor others
        % Their C counterparts are found in *_transport.h
        TRANSPORT_TYPE            = 0;
        TRANSPORT_HW_ADDR         = 1;
        TRANSPORT_IP_ADDR         = 2;
        TRANSPORT_UNICAST_PORT    = 3;
        TRANSPORT_BCAST_PORT      = 4;
        TRANSPORT_GRP_ID          = 5;
    end
    
    
%********************************* Methods ************************************

    methods
        function obj = wn_transport_eth_udp()
            obj.hdr         = wn_transport_header;
            obj.hdr.pktType = obj.hdr.PKTTYPE_HTON_MSG;
            obj.status      = 0;
            obj.timeout     = 1;
            obj.maxPayload  = 1000; % Sane default. This gets overwritten by CMD_PAYLOADSIZETEST command.
            
            obj.checkSetup();
        end
        
        function checkSetup(obj)
            %TODO
        end

        function setMaxPayload(obj,value)
            obj.maxPayload = value;
        end
        
        function out = getMaxPayload(obj)
            out = obj.maxPayload;
        end

        function open(obj,varargin)
            % Must be implemented by child classes
            obj.status = 0;
        end
        
        function out = procCmd(obj,nodeInd,node,cmdStr,varargin)
            out = [];
            cmdStr = lower(cmdStr);
            switch(cmdStr)
                case 'ping'
                    % Test to make sure node can be accessed via this
                    % transport
                    %
                    % Arguments: none
                    % Returns: true if board responds; raises error otherwise
                    myCmd = wn_cmd(node.calcCmd(obj.GRP,obj.CMD_PING));
                    node.sendCmd(myCmd);
                    out = true; %sendCmd will timeout and raise error if board doesn't respond
                    
                case 'payload_size_test'
                    % Determine's objects maxPayload parameter
                    %
                    % Arguments: none
                    % Returns: none
                    %
                    configFile = which('wn_config.ini');
                    if(isempty(configFile))
                       error(generatemsgid('MissingConfig'),'cannot find wn_config.ini. please run wn_setup.m'); 
                    end
                    readKeys = {'network','','jumbo',''};
                    jumbo = inifile(configFile,'read',readKeys);
                    jumbo = jumbo{1};
                    
                    if(strcmp('true',jumbo))
                        payloadTestSizes = [1000 1470 5000 8966];    
                    else
                        payloadTestSizes = [1000 1470];    
                    end
                    
                    payloadTestSizes = floor((payloadTestSizes - ( sizeof(node.transport.hdr) + sizeof(wn_cmd) ) - 4)/4);
                         
                    for index = 1:length(payloadTestSizes)
                        myCmd = wn_cmd(node.calcCmd(obj.GRP,obj.CMD_PAYLOADSIZETEST));
                        myCmd.addArgs(1:payloadTestSizes(index));
                        try
                            resp = node.sendCmd(myCmd);
                            obj.setMaxPayload( resp.getArgs );
                            if(obj.getMaxPayload < payloadTestSizes(index)*4)
                               break; 
                            end
                        catch ME
                            break 
                        end
                    end
                    
                case 'get_warpnet_type'
                    % Gets the WARPNet type from a node
                    %
                    % Arguments: none
                    % Returns:   type
                    %
                    myCmd = wn_cmd( node.calcCmd( node.GRP_WARPNET, node.CMD_WARPNET_TYPE ) );
                    resp  = node.sendCmd(myCmd);
                    out   = resp.getArgs();
                
                case 'add_node_group_id'
                    % Adds a Node Group ID to the node so that it can
                    % process broadcast commands that are received from
                    % that node group.
                    %
                    % Arguments: (uint32 NODE_GRP_ID)
                    % Returns: none
                    %
                    % NODE_GRP_ID: ID provided by wn_node_grp
                    %                   
                    NODE_GRP_ID = varargin{1};
                    myCmd = wn_cmd(node.calcCmd(obj.GRP,obj.CMD_NODEGRPID_ADD));
                    myCmd.addArgs(NODE_GRP_ID);
                    node.sendCmd(myCmd);
                    
                case 'clear_node_group_id'
                    % Clears a Node Group ID from the node so it can ignore
                    % broadcast commands that are received from that node
                    % group.
                    %
                    % Arguments: (uint32 NODE_GRP_ID)
                    % Returns: none
                    %
                    % NODE_GRP_ID: ID provided by wn_node_grp
                    %
                    NODE_GRP_ID = varargin{1};
                    myCmd = wn_cmd(node.calcCmd(obj.GRP,obj.CMD_NODEGRPID_CLEAR));
                    myCmd.addArgs(NODE_GRP_ID);
                    node.sendCmd(myCmd);
                    
                otherwise
                    error(generatemsgid('UnknownCmd'),'unknown command ''%s''',cmdStr);
            end
            
            if(iscell(out)==0 && numel(out)~=1)
                out = {out}; 
            end     
        end
        
        function close(obj)
            obj.status = 0;
        end    

        function delete(obj)
            obj.close();
        end

        function flush(obj)
            %TODO
        end
    end %methods

    methods (Hidden = true)
        function reply = send(obj,varargin)
            % Must be implemented by the child class
            reply = [];
        end
        
        function resp = receive(obj)
            % Must be implemented by the child class
            resp = [];
        end

        
        % Process parameters
        function processParameter( obj, identifier, length, values )        
            switch( identifier ) 
                case obj.TRANSPORT_TYPE
                    if ( length == 1 ) 
                        obj.type = values(1);
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport Type Parameter had incorrect length.  Should be 1, received %d', length);
                    end
                
                case obj.TRANSPORT_HW_ADDR
                    if ( length == 2 ) 
                        obj.eth_MAC_addr = 2^32*double(bitand(values(1),2^16-1)) + double(values(2));
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport HW Address Parameter had incorrect length.  Should be 2, received %d', length);
                    end
                    
                case obj.TRANSPORT_IP_ADDR
                    if ( length == 1 )
                        obj.address = obj.int2IP( values(1) );
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport IP Address Parameter had incorrect length.  Should be 1, received %d', length);
                    end
                    
                case obj.TRANSPORT_UNICAST_PORT                    
                    if ( length == 1 )
                        obj.port = values(1);
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport Unicast Port Parameter had incorrect length.  Should be 1, received %d', length);
                    end
                    
                case obj.TRANSPORT_BCAST_PORT                    
                    if ( length == 1 ) 
                        obj.bcast_port = values(1);
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport Broadcast Port Parameter had incorrect length.  Should be 1, received %d', length);
                    end
                    
                case obj.TRANSPORT_GRP_ID
                    if ( length == 1 )
                        group_id = values(1);
                    else
                        error(generatemsgid('IncorrectParameter'), 'Transport Group ID Parameter had incorrect length.  Should be 1, received %d', length);
                    end

                otherwise
                    error(generatemsgid('UnknownParameter'), 'Unknown Transport Group Parameter Identifier: %d', identifier);
            end
            
        end
        
                
        function dottedIPout = int2IP(obj,intIn)
            addrChars(4) = mod(intIn, 2^8);
            addrChars(3) = mod(bitshift(intIn, -8), 2^8);
            addrChars(2) = mod(bitshift(intIn, -16), 2^8);
            addrChars(1) = mod(bitshift(intIn, -24), 2^8);
            dottedIPout = sprintf('%d.%d.%d.%d', addrChars);
        end
        

        function intOut = IP2int(obj,dottedIP)
            addrChars = sscanf(dottedIP, '%d.%d.%d.%d')';
            intOut = 2^0 * addrChars(4) + 2^8 * addrChars(3) + 2^16 * addrChars(2) + 2^24 * addrChars(1);
        end
    end
end %classdef

File: ReferenceDesigns/w3_802.11/python/warpnet/wn_transport_eth_udp.py
# -*- coding: utf-8 -*-
"""
------------------------------------------------------------------------------
WARPNet Transport - Ethernet UDP base class
------------------------------------------------------------------------------
Authors:   Chris Hunter (chunter [at] mangocomm.com)
           Patrick Murphy (murphpo [at] mangocomm.com)
           Erik Welsh (welsh [at] mangocomm.com)
License:   Copyright 2014, Mango Communications. All rights reserved.
           Distributed under the WARP license (http://warpproject.org/license)
------------------------------------------------------------------------------
MODIFICATION HISTORY:

Ver   Who  Date     Changes
----- ---- -------- -----------------------------------------------------
1.00a ejw  1/23/14  Initial release

------------------------------------------------------------------------------

This module provides the base class for WARPNet Ethernet UDP transports.

Functions:
    WnTransportEthUdp() -- Base class for Ethernet UDP transports
    int2ip() -- Convert 32 bit integer to 'w.x.y.z' IP address string
    ip2int() -- Convert 'w.x.y.z' IP address string to 32 bit integer
    mac2str() -- Convert 6 byte MAC address to 'uu:vv:ww:xx:yy:zz' string

"""

import re
import time
import errno
import socket
from socket import error as socket_error

from . import wn_cmds
from . import wn_message
from . import wn_exception as ex
from . import wn_transport as tp


__all__ = ['WnTransportEthUdp']


class WnTransportEthUdp(tp.WnTransport):
    """Base Class for WARPNet Ethernet UDP Transport class.
       
    Attributes:
        timeout-- Maximum time spent waiting before retransmission
        transport_type -- Unique type of the WARPNet transport
        sock -- UDP socket
        status -- Status of the UDP socket
        max_payload -- Maximum payload size (eg. MTU - ETH/IP/UDP headers)
        
        mac_address -- Node's MAC address
        ip_address -- IP address of destination
        unicast_port -- Unicast port of destination
        bcast_port -- Broadcast port of destination
        group_id -- Group ID of the node attached to the transport
        rx_buffer_size -- OS's receive buffer size (in bytes)        
        tx_buffer_size -- OS's transmit buffer size (in bytes)        
    """
    timeout         = None
    transport_type  = None
    sock            = None
    status          = None
    max_payload     = None
    mac_address     = None
    ip_address      = None
    unicast_port    = None
    bcast_port      = None
    group_id        = None
    rx_buffer_size  = None
    tx_buffer_size  = None
    
    def __init__(self):
        self.hdr = wn_message.WnTransportHeader()
        self.status = 0
        self.timeout = 1
        self.max_payload = 1000   # Sane default.  Overwritten by payload test.
        
        self.check_setup()

    def __del__(self):
        """Closes any open sockets"""
        self.wn_close()

    def __repr__(self):
        """Return transport IP address and Ethernet MAC address"""
        return str("Eth UDP Transport: " + self.ip_address + "(" + self.eth_mac_addr + ")")

    def check_setup(self):
        """Check the setup of the transport."""
        pass

    def set_max_payload(self, value):  self.max_payload = value
    def set_ip_address(self, value):   self.ip_address = value
    def set_unicast_port(self, value): self.unicast_port = value
    def set_bcast_port(self, value):   self.bcast_port = value
    def set_src_id(self, value):       self.hdr.set_src_id(value)
    def set_dest_id(self, value):      self.hdr.set_dest_id(value)

    def get_max_payload(self):         return self.max_payload
    def get_ip_address(self):          return self.ip_address
    def get_unicast_port(self):        return self.unicast_port
    def get_bcast_port(self):          return self.bcast_port
    def get_src_id(self):              return self.hdr.get_src_id()
    def get_dest_id(self):             return self.hdr.get_dest_id()


    def wn_open(self, tx_buf_size=None, rx_buf_size=None):
        """Opens an Ethernet UDP socket.""" 
        self.sock = socket.socket(socket.AF_INET,       # Internet
                                  socket.SOCK_DGRAM);   # UDP
        
        self.sock.settimeout(self.timeout)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        
        if not tx_buf_size is None:
            try:
                self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, tx_buf_size)
            except socket_error as serr:
                # On some HW we cannot set the buffer size
                if serr.errno != errno.ENOBUFS:
                    raise serr

        if not rx_buf_size is None:
            try:
                self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, rx_buf_size)
            except socket_error as serr:
                # On some HW we cannot set the buffer size
                if serr.errno != errno.ENOBUFS:
                    raise serr

        self.tx_buffer_size = self.sock.getsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF)
        self.rx_buffer_size = self.sock.getsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF)
        self.status = 1
        
        if (self.tx_buffer_size != tx_buf_size):
            print("OS reduced send buffer size from {0} to {1}".format(self.tx_buffer_size, tx_buf_size))
            
        if (self.rx_buffer_size != rx_buf_size):
            print("OS reduced receive buffer size from {0} to {1}".format(self.rx_buffer_size, rx_buf_size))


    def wn_close(self):
        """Closes an Ethernet UDP socket."""
        if self.sock:
            try:
                self.sock.close()
            except socket.error as err:
                print("Error closing socket:  {0}".format(err))

        self.status = 0


    #-------------------------------------------------------------------------
    # Commands that must be implemented by child classes
    #-------------------------------------------------------------------------
    def send(self,):
        """Send a message over the transport."""
        raise NotImplementedError

    def receive(self,):
        """Return a response from the transport."""
        raise NotImplementedError


    #-------------------------------------------------------------------------
    # WARPNet Commands for the Transport
    #-------------------------------------------------------------------------
    def ping(self, node, output=False):
        """Issues a ping command to the given node.
        
        Will optionally print the result of the ping.
        """
        start_time = time.clock()
        node.send_cmd(wn_cmds.WnCmdPing())
        end_time = time.clock()
        
        if output:
            ms_time = (end_time - start_time) * 1000
            print("Reply from {0}:  time = {1:.3f} ms".format(self.ip_address, 
                                                              ms_time))
    
    def test_payload_size(self, node, jumbo_frame_support=False):
        """Determines the object's max_payload parameter."""

        if (jumbo_frame_support == True):
            payload_test_sizes = [1000, 1470, 5000, 8966]
        else:
            payload_test_sizes = [1000, 1470]
        
        for size in payload_test_sizes:
            my_arg_size = (size - (self.hdr.sizeof() + wn_message.WnCmd().sizeof() + 4)) // 4
 
            try:
                resp = node.send_cmd(wn_cmds.WnCmdTestPayloadSize(my_arg_size))
                self.set_max_payload(resp)
                
                if (self.get_max_payload() < (4*my_arg_size)):
                    # If the node received a smaller payload, then break
                    break
            except:
                # If there was a transmission error, then break
                break
    
    def add_node_group_id(self, node, group):
        node.send_cmd(wn_cmds.WnCmdAddNodeGrpId(group))
    
    def clear_node_group_id(self, node, group):
        node.send_cmd(wn_cmds.WnCmdClearNodeGrpId(group))


    #-------------------------------------------------------------------------
    # WARPNet Parameter Framework
    #   Allows for processing of hardware parameters
    #-------------------------------------------------------------------------
    def process_parameter(self, identifier, length, values):
        """Extract values from the parameters"""
        if   (identifier == tp.TRANSPORT_TYPE):
            if (length == 1):
                self.transport_type = values[0]
            else:
                raise ex.WnParameterError("TRANSPORT_TYPE", "Incorrect length")
                
        elif (identifier == tp.TRANSPORT_HW_ADDR):
            if (length == 2):
                self.mac_address = ((2**32) * (values[0] & 0xFFFF) + values[1])
            else:
                raise ex.WnParameterError("TRANSPORT_HW_ADDR", "Incorrect length")
                
        elif (identifier == tp.TRANSPORT_IP_ADDR):
            if (length == 1):
                self.ip_address = self.int2ip(values[0])
            else:
                raise ex.WnParameterError("TRANSPORT_IP_ADDR", "Incorrect length")
                
        elif (identifier == tp.TRANSPORT_UNICAST_PORT):
            if (length == 1):
                self.unicast_port = values[0]
            else:
                raise ex.WnParameterError("TRANSPORT_UNICAST_PORT", "Incorrect length")
                
        elif (identifier == tp.TRANSPORT_BCAST_PORT):
            if (length == 1):
                self.bcast_port = values[0]
            else:
                raise ex.WnParameterError("TRANSPORT_BCAST_PORT", "Incorrect length")
                
        elif (identifier == tp.TRANSPORT_GRP_ID):
            if (length == 1):
                self.group_id = values[0]
            else:
                raise ex.WnParameterError("TRANSPORT_GRP_ID", "Incorrect length")
                
        else:
            raise ex.WnParameterError(identifier, "Unknown transport parameter")


    #-------------------------------------------------------------------------
    # Misc methods for the Transport
    #-------------------------------------------------------------------------
    def int2ip(self, ipaddr):  return int2ip(ipaddr)

    def ip2int(self, ipaddr):  return ip2int(ipaddr)

    def mac2str(self, mac_address):  return mac2str(mac_address)

    def __str__(self):
        """Pretty print the Transport parameters"""
        print("Transport {}:".format(self.transport_type))
        print("    IP address    :  {}".format(self.ip_address))
        print("    MAC address   :  {}".format(self.mac2str(self.mac_address)))
        print("    Max payload   :  {}".format(self.max_payload))
        print("    Unicast port  :  {}".format(self.unicast_port))
        print("    Broadcast port:  {}".format(self.bcast_port))
        print("    Timeout       :  {}".format(self.timeout))
        print("    Rx Buffer Size:  {}".format(self.rx_buffer_size))
        print("    Tx Buffer Size:  {}".format(self.tx_buffer_size))
        print("    Group ID      :  {}".format(self.group_id))
        
# End Class WnTransportEthUdp



#-------------------------------------------------------------------------
# Global methods for the Transport
#-------------------------------------------------------------------------

def int2ip(ipaddr):
    return str("{0:d}.{1:d}.{2:d}.{3:d}".format(((ipaddr >> 24) & 0xFF),
                                                ((ipaddr >> 16) & 0xFF),
                                                ((ipaddr >>  8) & 0xFF),
                                                (ipaddr & 0xFF)))


def ip2int(ipaddr):
    expr = re.compile('\.')
    dataTuple = []
    for data in expr.split(ipaddr):
        dataTuple.append(int(data))
    return (dataTuple[3]) + (dataTuple[2] * 2**8) + (dataTuple[1] * 2**16) + (dataTuple[0] * 2**24)


def mac2str(mac_address):
    return str("{0:02x}".format((mac_address >> 40) & 0xFF) + ":" +
               "{0:02x}".format((mac_address >> 32) & 0xFF) + ":" +
               "{0:02x}".format((mac_address >> 24) & 0xFF) + ":" +
               "{0:02x}".format((mac_address >> 16) & 0xFF) + ":" +
               "{0:02x}".format((mac_address >>  8) & 0xFF) + ":" +
               "{0:02x}".format(mac_address & 0xFF))









